### PR TITLE
Add enchancement: Configurable Image Versions

### PR DESCRIPTION
**Pull request by lranjbar**
# Assisted Pull Request

## Description

Enhancement about image versions in the code.

## List all the issues related to this PR

- [X] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [X] Cloud
- [X] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @mhrivnak 
/cc @flaper87 

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by lranjbar** [comment]
Updated with notes from the architecture meeting from yesterday.

**Review by mhrivnak — CHANGES_REQUESTED** [review]
Looks good. My comments are mostly about clarifying some points.

File: docs/enhancements/config-image-versions.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+with each other, I ask the team to rethink this. The version of these images to run 
+is determined at runtime by the user. When writing the code we do not know this
+value.
+
+By supplying default values we run the risk of trying to run components that are 
+incompatible with each other.
+
+### Open Questions
+
+- Is the reason these default image values are supplied because it is used in code
```
[the remark this answers is not in the document]
Comment (lranjbar) [reply docs/enhancements/config-image-versions.md:98]: The consensus on this was mostly that these were supplied for convenience and not linked to logic in the code.

File: docs/enhancements/config-image-versions.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+in the ConfigMap at runtime.
+
+Second, have errors thrown when the environment variables for the images are 
+blank and not provided in the ConfigMap. If a user doesn't supply a value for a 
+required image we should give a runtime error for the user.
+
+Third, throw warnings (or errors) when the version in the environment variable 
+differs from what is supplied at runtime for Assisted Installer.
+
+Finally, it might be good to validate that all the version tags of Assisted
```
[the remark this answers is not in the document]
Comment (lranjbar) [reply docs/enhancements/config-image-versions.md:57]: Right this is probably a warning and less an error because there are plenty of valid reasons to update one out of the set at deployment time.

File: docs/enhancements/config-image-versions.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+First, if we need to populate these environment variables in the code then 
+Assisted Installer should populate them with the environment variables given 
+in the ConfigMap at runtime.
+
+Second, have errors thrown when the environment variables for the images are 
+blank and not provided in the ConfigMap. If a user doesn't supply a value for a 
+required image we should give a runtime error for the user.
+
+Third, throw warnings (or errors) when the version in the environment variable
```
[the remark this answers is not in the document]
Comment (lranjbar) [reply docs/enhancements/config-image-versions.md:54]: Maybe the confusion is that mostly you have to define these environment variables at runtime for things to work properly. Right now these environment variables have default values though in the code, so if you don't define them at runtime it falls back to those values. What I'm trying to say is we need to error if those environment variables are null or empty. These values for the environment variables should be defined by the user in the artifacts needed for deployment. 

These values of the environment variables could change at runtime from our perspective as the Assisted Installer development team. Another way of wording this is we can't assume in our code which version of Assisted Installer a user will choose to deploy and run.

As a result these values actually do fall in the ConfigMaps and other deployment artifacts. This is kind of what Igal was getting at in the meeting. These things are defined in many places, having a default value in the code kind of puts it in "one" place if you want to use that version. 

The problem is when you want to change the version of the images you are using you have to define it all those different places anyway. The default values technically only hurt you in that case, because it is very unclear what is being deployed. The deployment itself doesn't throw an error in Openshift for you to debug. In fact, Assisted Installer didn't error at all when the deployment method wasn't set up correctly. It just ran the incorrect version of Assisted Installer.

[the remark this answers is not in the document]
Comment (lranjbar) [reply docs/enhancements/config-image-versions.md:54]: I hope it clarifies.. but maybe it doesn't.

File: docs/enhancements/config-image-versions.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+There are default versions of Openshift Container Platform and RHCOS images
+supplied to Assisted Installer. These are not the focus of this discussion only
+how we decide which images to deploy and interact with.
+
+## Proposal
+
+First, if we need to populate these environment variables in the code then 
+Assisted Installer should populate them with the environment variables given 
+in the ConfigMap at runtime.
```
[the remark this answers is not in the document]
Comment (mhrivnak) [reply docs/enhancements/config-image-versions.md:48]: For the SaaS, or generally non-operator deployments on k8s, I wouldn't worry about explaining any automation that exists. Probably what matters is this, assuming I have the facts right: Environment variables need to be set on the assisted-service Pod, telling it what container images to use, and that is typically accomplished by putting those values in a ConfigMap and referencing the ConfigMap in the Pod definition.

For the operator: it looks like the CSV's Deployment definition has [image references hard-coded as environment variables](https://github.com/openshift/assisted-service/blob/bff1eb2/deploy/olm-catalog/manifests/assisted-service-operator.clusterserviceversion.yaml#L579-L591). Then I assume when the operator deploys the operand, it puts those same values into the same environment variables in a ConfigMap, and references that from the assisted-service Deployment definition? Or does it skip the ConfigMap and hard-code them onto the Deployment definition?

So I think there are two ways to take the above info and make the enhancement more clear:

1. Add the above details to be clear about where the configmap comes into play, and what the current expectation is at deployment time vs. the proposed expectation. There are several references in the proposal to "the ConfigMap", and it would help to clarify exactly what that is.
2. Don't mention configmaps or other mechanisms at all. Instead, just focus on the environment variables as the API. Currently they are optional, and the proposal is for them to be mandatory. Someone or something will have to set them, and it doesn't matter if it's by configmap, in a Pod definition, or any other way.

I'm leaning toward 2. The configmap might just be an implementation detail that isn't central to this proposal. WDYT?

File: docs/enhancements/config-image-versions.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mhrivnak) [reply docs/enhancements/config-image-versions.md:54]: Your comment makes sense. But it seems like what you're describing is covered by the paragraph above: "Second, have errors...". It's not clear to me what this third scenario is that's different from the second.

[the remark this answers is not in the document]
Comment (lranjbar) [reply docs/enhancements/config-image-versions.md:54]: Need to think on how I can reword it to be more clear. In essence the default values for these things would still exist, just in the deployment artifacts not in the code itself.

File: docs/enhancements/config-image-versions.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+#### Story 2
+
+As a user of Assisted Installer I expect to have basic validations on missing
+values in the ConfigMap.
+
+#### Story 3
+
+As a user of Assisted Installer I expect to be warned if my image configuration
+drifts during runtime. (This could be a sign of instability of my installation.)
```
[the remark this answers is not in the document]
Comment (lranjbar) [reply docs/enhancements/config-image-versions.md:76]: This is basically trying to detect if the value in these environment variables have changed after Assisted Installer is already running. I'm still trying to think of how exactly we might get in this state, but essentially it seems to me like it would be not great. 🤔

File: docs/enhancements/config-image-versions.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mhrivnak) [reply docs/enhancements/config-image-versions.md:54]: I think it's helpful to conceptually separate the software from any runtime configuration, regardless of where that config comes from. I would focus this proposal on the software, and just say we're removing defaults from the software.

As a next layer, various packaging and config management tools can be used to manage the runtime configuration. Those tools might or might not offer defaults with an option for user-supplied config. But that's all in the next layer up, which I don't think this proposal needs to address.

File: docs/enhancements/config-image-versions.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (lranjbar) [reply docs/enhancements/config-image-versions.md:48]: To me option 2 makes more sense for this proposal. It wasn't really the intention to try to codify how the different deployment methods would change in this enhancement. Obviously they would need to change but I feel these are implementation details that will need to be ironed out once the environment variables become required.
